### DISCUSSION TITLE
Mobile (Android/iOS/C++) bindings

### DISCUSSION THREAD

**Discussion by jsamol** [smol-dot/smoldot#1331] [All]
I'm looking into embedding Smoldot into a native mobile application. Has anyone done that already and have fully operational native (C++) bindings that could be used in an Android or iOS application?

**Comment by tomaka** [comment]
It is unclear to me (I couldn't find the answer on the Internet) whether, when embedding Rust within C/C++, it is possible for the Rust code to use the Rust standard library.

If the answer is yes, then the bindings should be relatively simple to write. If the answer is no, then one will need to implement the `PlatformRef` trait, which is a bit complicated by doable.

**Reply by tomaka** [reply]
It seems that yes, there's no problem with using the stdlib.

I've tried it out and wrote some quick C bindings: https://github.com/smol-dot/smoldot/compare/main...tomaka:smoldot:c-bindings

**Reply by jsamol** [reply]
Hi! Sorry it took me this long to respond 😅
 
First of all, thanks for looking into this! 😄 When I started working on it I wasn't aware of `smoldot_light::platform::DefaultPlatform` so I began to adapt [the C bindings](https://github.com/smol-dot/smoldot/blob/main/wasm-node/rust/src/bindings.rs)
that I found in the [`wasm-node` crate](https://github.com/smol-dot/smoldot/tree/main/wasm-node), hoping I could benefit from the stability and maturity of that code base.
 
Now, following your example bindings, I slimmed the code down by removing the custom `PlatformRef` implementation and replacing it with the default one. I also had to adapt some of the types in the bindings, mostly to add support for 64-bit environments: [jsamol/smoldot/mobile-node](https://github.com/jsamol/smoldot/tree/feat/mobile-bindings/mobile-node)

It did compile and I ran it successfully on an Android device. It appears to work fine as long as I use it with relay chains. However, when I add a parachain and send a JSON-RPC request, I don't see any responses received back (I haven't done proper debugging yet, but from what I've seen so far, [the stream just hangs on reading the next response](https://github.com/jsamol/smoldot/blob/feat/mobile-bindings/mobile-node/rust/src/lib.rs#L154)).

Before I dive deeper into this problem I wanted to check, is this somehow a known issue? Or maybe I missed something while adapting the wasm bindings?

I'd really appreciate any help that could put me on the right track while debugging this.

**Reply by tomaka** [reply]
The easiest way to figure out what's happening is to enable logging and copy-paste here what the debug logs say.

The light client uses the `log` crate (well, I've just opened https://github.com/smol-dot/smoldot/pull/1560, so it will no longer be the case in the very near future).

See https://github.com/smol-dot/smoldot/blob/fd23bfb6ae718f8e5ee153f4e10dbd6add9948b6/light-base/examples/basic.rs#L22-L25 or https://github.com/smol-dot/smoldot/blob/fd23bfb6ae718f8e5ee153f4e10dbd6add9948b6/wasm-node/rust/src/init.rs#L59 for example.

**Reply by jsamol** [reply]
Here are the logs: https://gist.github.com/jsamol/9c0a838398e1d1d2535d316b460d84a3

I tested the code with Kusama and Karura, whose specs I took from the [`demo-chain-specs` folder](https://github.com/smol-dot/smoldot/tree/main/demo-chain-specs). I called `chain_getBlockHash` on Kusama first and then the same method on Karura. Both actions are logged on lines [#145](https://gist.github.com/jsamol/9c0a838398e1d1d2535d316b460d84a3#file-smoldotmobile_080124_131427-log-L145) and [#147](https://gist.github.com/jsamol/9c0a838398e1d1d2535d316b460d84a3#file-smoldotmobile_080124_131427-log-L147) respectively. Only Kusama returns a response on line [#249](https://gist.github.com/jsamol/9c0a838398e1d1d2535d316b460d84a3#file-smoldotmobile_080124_131427-log-L249).

That's the only sensible information I was able to get, there are so many logs on the debug level 😅 I wasn't sure what's relevant so I just let it run for a minute, hoping I'd gather the data that matters

**Reply by tomaka** [reply]
Don't use the specs from the `demo-chain-specs` folder, they are for internal testing.

The problem here is that none of the bootnodes of the Karura chain spec are reachable.

**Reply by jsamol** [reply]
Oh, I see! It does work indeed after I switched to use the Karura specs found in their repo. Just for me to understand and, hopefully, be able to debug such issues better in future, which parts of the logs gave it away? I don't really see a clear message suggesting there might be connectivity issues.

**Reply by tomaka** [reply]
There isn't any message that will say that you're not connected, but there is the absence of a message that says that you're connected.

The plan to make this more clear is https://github.com/smol-dot/smoldot/issues/1343

**Reply by jsamol** [reply]
Ok, got it. Thanks for your help, it's very appreciated!

**Reply by jsamol** [reply]
@tomaka I'm coming with an update and more questions to ask 😅 

As I mentioned, I did eventually manage to write bindings that I could run on an Android device and the client, in overall, works. However, when I started using it more and more in my code, I noticed that the client's behaviour is not necessarily something I can rely on.

For example, when I try to call an extrinsic quite early after the app's launch and it's the first time the app is running (so no previously persisted state to continue from), the extrinsic usually fails to submit with the `BadProof` error. The same transaction passes through on the very next run, which leads me to believe the first attempt might be happening before the client is fully synced and uses outdated data to construct the extrinsic. Is it possible? If so, can I mitigate it somehow?

Other than that, I also get a lot of `No node available for storage query` and `Error trying to access the storage required for the runtime call` when, well, querying the storage, usually at the beginning of my app's lifecycle. Is it expected? Any advice what I could do about it?

**Reply by tomaka** [reply]
I have written https://forum.polkadot.network/t/new-json-rpc-api-mega-q-a/3048/17, which might concern your questions
